src/mlio/image_reader.cxx

```diff
@@ -0,0 +1,392 @@
+image_reader::make_record_reader(data_store const &ds)
+{
+    switch (params_.img_frame) {
+    case image_frame::none:
+        return make_intrusive<blob_record_reader>(ds.open_read());
+    case image_frame::recordio:
+        return make_intrusive<detail::recordio_record_reader>(ds.open_read());
+    }
+    throw std::invalid_argument{"Please enter a valid value for "
+                                "img_frame parameter."};
+}
+
+void
+image_reader::infer_schema(instance const &)
+{
+    std::vector<feature_desc> descs{};
+    // schema follows the NHWC convention
+    descs.emplace_back(feature_desc_builder{
+        "value",
+        data_type::uint8,
+        {params().batch_size,
+         static_cast<std::size_t>(params_.image_dimensions[1]),
+         static_cast<std::size_t>(params_.image_dimensions[2]),
+         static_cast<std::size_t>(params_.image_dimensions[0])}}
+                           .build());
+    schema_ = make_intrusive<schema>(std::move(descs));
+}
+
+intrusive_ptr<example>
+image_reader::decode(instance_batch const &batch) const
+{
+    auto tsr = make_tensor(batch.instances(), batch.size());
+    if (tsr == nullptr) {
+        return nullptr;
+    }
+    std::vector<intrusive_ptr<tensor>> tensors{};
+    tensors.emplace_back(std::move(tsr));
+    return make_intrusive<example>(schema_, std::move(tensors));
+}
+
+intrusive_ptr<dense_tensor>
+image_reader::make_tensor(std::vector<instance> const &instances,
+                          std::size_t batch_size) const
+{
+    std::atomic_bool skip_batch{};
+    auto instance_offset = params_.image_dimensions[0] *
+                           params_.image_dimensions[1] *
+                           params_.image_dimensions[2];
+    auto total_memory_buffer =
+        static_cast<unsigned long>(instance_offset) * batch_size;
+    std::vector<std::uint8_t> img_vec(total_memory_buffer);
+    auto *ptr_img_vec = img_vec.data();
+    int type;
+    cv::ImreadModes mode;
+    for (instance const &ins : instances) {
+        memory_slice img_buf;
+        if (params_.img_frame == image_frame::recordio) {
+            img_buf = ins.bits().subslice(recordio_image_header_offset_);
+        }
+        else {
+            img_buf = ins.bits();
+        }
+        cv::Mat mat(
+            1,
+            static_cast<int>(img_buf.size()),
+            CV_8U,
+            static_cast<void *>(const_cast<std::byte *>(img_buf.data())));
+        switch (params_.image_dimensions[0]) {
+        case 1:
+            mode = cv::ImreadModes::IMREAD_GRAYSCALE;
+            type = CV_8UC1;
+            break;
+        case 3:
+            mode = cv::ImreadModes::IMREAD_COLOR;
+            type = CV_8UC3;
+            break;
+        case 4:
+            mode = cv::ImreadModes::IMREAD_UNCHANGED;
+            type = CV_8UC4;
+            break;
+        default:
+            throw std::invalid_argument{
+                fmt::format("Unsupported number of channels: {0:n}",
```

Comment:
So this will be a user error. User would have passed an invalid argument for the `image_dimensions_` parameter. That is why it wouldnt make sense for the error message to talk about the `data_store` or `index`.

Reply:
Opinion?